--- a/arch/arm64/boot/dts/mediatek/mt7988a-bananapi-bpi-r4-pro.dtsi
+++ b/arch/arm64/boot/dts/mediatek/mt7988a-bananapi-bpi-r4-pro.dtsi
@@ -436,6 +436,13 @@
 				status = "okay";
 			};
 
+			port4: port@4 {
+				reg = <4>;
+				label = "lan5";
+				phy-handle = <&switchphy4>;
+				phy-mode = "internal";
+				status = "okay";
+			};
 			port8: port@8 {
 				reg = <8>;
 				label = "cpu";
--- a/arch/arm64/boot/dts/mediatek/mt7988a-bananapi-bpi-r4-pro-8x.dts
+++ b/arch/arm64/boot/dts/mediatek/mt7988a-bananapi-bpi-r4-pro-8x.dts
@@ -43,6 +43,31 @@
 				sfp = <&sfp2>;
 			};
 		};
+
+		mux2: ethernet-mux@2 {
+			compatible = "mediatek,eth-mux";
+			reg = <2>; /* map to gmac2 */
+			chan-sel-gpios = <&pio 54 GPIO_ACTIVE_HIGH>;
+			mod-def0-gpios = <&pio 69 GPIO_ACTIVE_LOW>;
+			sfp-present-channel = <0>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			channel_20: channel@0 {
+				reg = <0>;
+				phy-mode = "10gbase-r";
+				phy-connection-type = "10gbase-r";
+				managed = "in-band-status";
+				sfp = <&sfp1>;
+			};
+
+			channel_21: channel@1 {
+				reg = <1>;
+				phy-mode = "10gbase-r";
+				phy-connection-type = "10gbase-r";
+				phy-handle = <&phy24>;
+			};
+		};
 	};
 };
 
@@ -145,7 +170,7 @@
 		};
 	};
 	ports {
-		port6: port@6 {
+		port12: port@12 {
 			reg = <12>;
 			label = "lan4";
 			phy-mode = "usxgmii";
