--- a/arch/arm/dts/mt7988.dtsi
+++ b/arch/arm/dts/mt7988.dtsi
@@ -412,7 +412,7 @@
 	};
 
 	uart0: serial@11000000 {
-		compatible = "mediatek,hsuart";
+		compatible = "mediatek,hsuart", "mediatek,mt6577-uart";
 		reg = <0 0x11000000 0 0x100>;
 		interrupts = <GIC_SPI 123 IRQ_TYPE_LEVEL_HIGH>;
 		clocks = <&infracfg CLK_INFRA_52M_UART0_CK>;
@@ -424,7 +424,7 @@
 	};
 
 	uart1: serial@11000100 {
-		compatible = "mediatek,hsuart";
+		compatible = "mediatek,hsuart", "mediatek,mt6577-uart";
 		reg = <0 0x11000100 0 0x100>;
 		interrupts = <GIC_SPI 124 IRQ_TYPE_LEVEL_HIGH>;
 		clocks = <&infracfg CLK_INFRA_52M_UART1_CK>;
@@ -436,7 +436,7 @@
 	};
 
 	uart2: serial@11000200 {
-		compatible = "mediatek,hsuart";
+		compatible = "mediatek,hsuart", "mediatek,mt6577-uart";
 		reg = <0 0x11000200 0 0x100>;
 		interrupts = <GIC_SPI 125 IRQ_TYPE_LEVEL_HIGH>;
 		clocks = <&infracfg CLK_INFRA_52M_UART2_CK>;
