--- a/boot/image-fit.c
+++ b/boot/image-fit.c
@@ -2150,7 +2150,7 @@ int fit_image_load(struct bootm_headers
 		if (ret < 0 && ret != -EINVAL)
 			ret = fit_conf_get_node(fit, fit_uname_config);
 		if (ret < 0) {
-			puts("Could not find configuration node\n");
+			printf("Could not find configuration node '%s'\n", fit_uname_config);
 			bootstage_error(bootstage_id +
 					BOOTSTAGE_SUB_NO_UNIT_NAME);
 			return -ENOENT;
